node_modules/col-assets/lib/api.js

```diff
@@ -395,35 +403,53 @@ var getAssets = module.exports.getAssets = function(ctx, filters, sort, limit, o
     sqlQuery += hasImpact ? ' AND a.impact_score > 0' : ' AND a.impact_score = 0';
   }
 
-  var hasTrending = CollabosphereUtil.getBooleanParam(filters.hasTrending, null);
-  if (hasTrending !== null) {
-    sqlQuery += hasTrending ? ' AND a.trending_score > 0' : ' AND a.trending_score = 0';
-  }
-
   var hasLikes = CollabosphereUtil.getBooleanParam(filters.hasLikes, null);
   if (hasLikes !== null) {
     sqlQuery += hasLikes ? ' AND a.likes > 0' : ' AND a.likes = 0';
   }
 
+  var hasPins = CollabosphereUtil.getBooleanParam(filters.hasPins, null);
+  if (hasPins !== null) {
+    sqlQuery += hasPins ? ' AND p.asset_id IS NOT NULL' : ' AND p.asset_id IS NULL';
+  }
+
+  var hasTrending = CollabosphereUtil.getBooleanParam(filters.hasTrending, null);
+  if (hasTrending !== null) {
+    sqlQuery += hasTrending ? ' AND a.trending_score > 0' : ' AND a.trending_score = 0';
+  }
+
   var hasViews = CollabosphereUtil.getBooleanParam(filters.hasViews, null);
   if (hasViews !== null) {
     sqlQuery += hasViews ? ' AND a.views > 0' : ' AND a.views = 0';
   }
 
   // Query that will be used to get the actual assets
-  var resultsSqlQuery = 'SELECT DISTINCT ON (a.id, a.likes, a.views, a.comment_count, a.impact_score, a.trending_score) a.*, act.type AS activity_type ' + sqlQuery;
-  if (sort === 'recent') {
-    resultsSqlQuery += ' ORDER BY a.id DESC';
-  } else if (sort === 'likes') {
-    resultsSqlQuery += ' ORDER BY a.likes DESC, a.id DESC';
-  } else if (sort === 'views') {
-    resultsSqlQuery += ' ORDER BY a.views DESC, a.id DESC';
-  } else if (sort === 'comments') {
-    resultsSqlQuery += ' ORDER BY a.comment_count DESC, a.id DESC';
-  } else if (sort === 'impact') {
-    resultsSqlQuery += ' ORDER BY a.impact_score DESC, a.id DESC';
-  } else if (sort === 'trending') {
-    resultsSqlQuery += ' ORDER BY a.trending_score DESC, a.id DESC';
+  var resultsSqlQuery = `SELECT
+    DISTINCT ON (a.id, a.likes, a.views, a.comment_count, a.impact_score, a.trending_score)
+    a.*,
+    act.type AS activity_type,
+    (select count(*) from pinned_user_assets p where p.asset_id = a.id)::int AS pin_count
+  `;
+  resultsSqlQuery += sqlQuery;
+
+  // A nasty SQL 'GROUP BY' clause is needed to sort by `pin_count`.
+  // Instead, sort by pin_count after query execution.
+  var sortByPinned = (sort === 'pins');
```

Comment:
@johncrossman I think the problem with sorting after query execution is that it won't play nicely with the LIMIT and OFFSET that we use to retrieve assets in groups of 10. I'd expect this to sort assets by pin_count within each group of 10, but the groups of 10 themselves would come back without explicit ordering--which might even do something odd like return the oldest groups first. Am I missing something? If not, the choice might be between a) nasty SQL and b) not offering pin_count as a sort option. I think b) is honorable.